### ISSUE TITLE
[java] MissingStaticMethodInNonInstantiatableClass does not consider nested builder class

### ISSUE THREAD

**Issue by mihalyr** [pmd/pmd#4174]
**Affects PMD Version:** 6.50

**Rule:** [MissingStaticMethodInNonInstantiatableClass](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#missingstaticmethodinnoninstantiatableclass)

**Description:** A private final class can be instantiated using a nested static class which might return an interface with the class as the implementation. I think the problem is that when an interface is returned PMD does not understand that it is actually returning an instance of the same class.

**Code Sample demonstrating the issue:**

file: `GenericTest.java`

```java
public interface GenericTest {
  void run();
}
```

file: `TestPrivateClassWithFactory.java`

```java
public final class TestPrivateClassWithFactory implements GenericTest {
  private final boolean shouldFail;

  private TestPrivateClassWithFactory(TestStep step) {
    this.shouldFail = step.shouldFail;
  }

  public void run() {
    if (shouldFail) {
      throw new AssertionError("fail");
    }
  }

  public static final class TestStep {
    private boolean shouldFail = false;

    public TestStep shouldFail() {
      this.shouldFail = true;
      return this;
    }

    public GenericTest newTest() {
      return new TestPrivateClassWithFactory(this);
    }
  }
}
```

**Expected outcome:** PMD does not complain about `MissingStaticMethodInNonInstantiatableClass`

PMD reports a violation at line `1` column `14` in `TestPrivateClassWithFactory.java` but that's wrong. That's a false positive.

**Running PMD through:** *Maven and also IntelliJ's PMD-IDEA plugin*
